### PR TITLE
Issue #3263200 by vnech: As a group manager I want to be able to send a reminder about invitation to the group

### PR DESCRIPTION
**Pull request by nechai**
## Problem
Sometimes group managers needs to resend membership invitation (as users can forgot to see the previous one or it can be "lost" in the mess of notifications or emails).

Just using the default "invitation sending" behaviour doesn't work because users doesn't receive the invitation again (as OS doesn't allow to create invitation duplicate).

## Solution

- Add new action "Send reminder" to each invite on the "Invitations" page
- Add bulk operations (same on "Manage members") for invites on the "Invitation" page
- Add UI for SM to define resend invitation message instead default one
- Add "resend" invitation feature

In general, "Send reminder" operation is only deleting old invitation and create a new one.

## Issue tracker
- https://www.drupal.org/project/social/issues/3263200
- https://getopensocial.atlassian.net/browse/YANG-6005

## How to test
- [x] Login as CM+
- [x] Create a flexible group
- [x] Make sure current users have a role "Group manager" for the created group
- [x] Invite a few users to a group - _/group/{gid}/invite-members_
- [x] Send reminders on the page - _/group/{gid}/invitation-overview_

## Screenshots
Screenshot 2022-03-10 at 09 21 32

## Release notes
*Allow group managers to resend membership invitations*

## Change Record
N/A

## Translations
N/A

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by Ressinel** [comment]
@nechai Could you fill the HTT section, please?

**Review by Ressinel — COMMENTED** [review]
@nechai As GM I can't send a reminder by bulk operation:
- Sent two invites;
- Got to invites list;
- Select all invites and select action **Send reminders**;
- You will be redirected to page `/group/[group_id]/invites/confirm-action` and see that you have no access to that page `You are not authorized to access this page.` with next waring

```
Notice: Undefined index: view_id in Drupal\views_bulk_operations\Access\ViewsBulkOperationsAccess->access() (line 42 of modules/contrib/views_bulk_operations/src/Access/ViewsBulkOperationsAccess.php).

Drupal\views_bulk_operations\Access\ViewsBulkOperationsAccess->access(Object, Object) (Line: 28)
Drupal\social_group_invite\Access\SocialGroupInviteVBOAccess->access(Object, Object)
call_user_func_array(Array, Array) (Line: 159)
Drupal\Core\Access\AccessManager->performCheck('social_group_invite.views_bulk_operations.access', Object) (Line: 135)
Drupal\Core\Access\AccessManager->check(Object, Object, Object, 1) (Line: 112)
Drupal\Core\Access\AccessManager->checkRequest(Object, Object, 1) (Line: 109)
Drupal\Core\Routing\AccessAwareRouter->checkAccess(Object) (Line: 94)
Drupal\Core\Routing\AccessAwareRouter->matchRequest(Object) (Line: 112)
Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(Object, 'kernel.request', Object)
call_user_func(Array, Object, 'kernel.request', Object) (Line: 142)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch(Object, 'kernel.request') (Line: 134)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 80)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 717)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```

```
Warning: array_flip(): Can only flip STRING and INTEGER values! in Drupal\Core\Entity\EntityStorageBase->loadMultiple() (line 261 of core/lib/Drupal/Core/Entity/EntityStorageBase.php).

Drupal\Core\Entity\EntityStorageBase->loadMultiple(Array) (Line: 245)
Drupal\Core\Entity\EntityStorageBase->load(NULL) (Line: 122)
Drupal\views\Views::getView(NULL) (Line: 42)
Drupal\views_bulk_operations\Access\ViewsBulkOperationsAccess->access(Object, Object) (Line: 28)
Drupal\social_group_invite\Access\SocialGroupInviteVBOAccess->access(Object, Object)
call_user_func_array(Array, Array) (Line: 159)
Drupal\Core\Access\AccessManager->performCheck('social_group_invite.views_bulk_operations.access', Object) (Line: 135)
Drupal\Core\Access\AccessManager->check(Object, Object, Object, 1) (Line: 112)
Drupal\Core\Access\AccessManager->checkRequest(Object, Object, 1) (Line: 109)
Drupal\Core\Routing\AccessAwareRouter->checkAccess(Object) (Line: 94)
Drupal\Core\Routing\AccessAwareRouter->matchRequest(Object) (Line: 112)
Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(Object, 'kernel.request', Object)
call_user_func(Array, Object, 'kernel.request', Object) (Line: 142)
Drupal\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch(Object, 'kernel.request') (Line: 134)
Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object, 1) (Line: 80)
Symfony\Component\HttpKernel\HttpKernel->handle(Object, 1, 1) (Line: 57)
Drupal\Core\StackMiddleware\Session->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object, 1, 1) (Line: 106)
Drupal\page_cache\StackMiddleware\PageCache->pass(Object, 1, 1) (Line: 85)
Drupal\page_cache\StackMiddleware\PageCache->handle(Object, 1, 1) (Line: 47)
Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object, 1, 1) (Line: 52)
Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object, 1, 1) (Line: 23)
Stack\StackedHttpKernel->handle(Object, 1, 1) (Line: 717)
Drupal\Core\DrupalKernel->handle(Object) (Line: 19)
```

**Review by Ressinel — COMMENTED** [review]
@nechai It would be nice to add behat test covering for these improvements.

**Comment by nechai** [comment]
@Ressinel Regarding [your comment ](https://github.com/goalgorilla/open_social/pull/2757#pullrequestreview-902792719)- I added fixes for accessing to "send reminder" operation. Make sure your user have a role "group manager "(you can find more details in HTT section)

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by Ressinel** [comment]
@nechai Fix conflict, please.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by nechai** [comment]
@Ressinel Conflicts fixed

File: phpstan-baseline.neon [hunk 1]
```diff
@@ -19709,3 +19704,43 @@ parameters:
 			count: 1
 			path: modules/social_features/social_group/src/Plugin/views/filter/FilterByGroup.php
 
+		-
```
Comment (Ressinel) [inline phpstan-baseline.neon:19707]: @nechai We should try to fix all possible PHPStan errors and not add new ones to the baseline file.

**Review by Ressinel — COMMENTED** [review]
@nechai Functionality works as espected. But I have added some suggestions for code. Take a look, please.

File: phpstan-baseline.neon (hunk 1 above)
Comment (nechai) [reply phpstan-baseline.neon:19707]: @Ressinel Fixed as much as possible.

File: modules/social_features/social_group/modules/social_group_invite/src/Form/SocialGroupInviteResendConfirmForm.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state): void {
+
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/src/Form/SocialGroupInviteResendConfirmForm.php:94]: Fixed.

File: modules/social_features/social_group/modules/social_group_invite/src/Form/SocialGroupInviteVBOConfirmAction.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Drupal\social_group_invite\Form;
+
+use Drupal\social_group_gvbo\Form\SocialGroupViewsBulkOperationsConfirmAction;
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Default action execution confirmation form.
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/src/Form/SocialGroupInviteVBOConfirmAction.php:9]: Fixed.

File: modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\TempStore\SharedTempStoreFactory;
+use Drupal\views_bulk_operations\Action\ViewsBulkOperationsActionBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Resend invites for group members.
+ *
+ *   This action allows to resend membership invitations to users. The logic
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php:15]: Fixed.

File: modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    $duplicate->setChangedTime($time);
+
+    // Set new invite to temp storage. This needs to identify if recipient of
+    // this invite should receive the altered invite message.
+    // Unfortunately, sending emails is triggering on entity inserting,
+    // and we don't have an "id" yet, so we should "uuid" property.
+    /* @see ginvite_group_content_insert() */
+    $values = (array) $this->tempStore->get(SocialGroupInviteResend::TEMP_STORE_ID);
+    $values[$duplicate->uuid()] = $duplicate->uuid();
+    $this->tempStore->set(SocialGroupInviteResend::TEMP_STORE_ID, $values);
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php:110]: Fixed.

File: modules/social_features/social_group/modules/social_group_invite/social_group_invite.module [hunk 6]
```diff
@@ -640,8 +665,29 @@ function social_group_invite_mail_alter(&$message) {
       !is_null($invite_settings) &&
       isset($invite_settings['invite_subject'], $invite_settings['invite_message'])
     ) {
+
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/social_group_invite.module:668]: Fixed.

File: modules/social_features/social_group/modules/social_group_invite/social_group_invite.module [hunk 7]
```diff
@@ -640,8 +665,29 @@ function social_group_invite_mail_alter(&$message) {
...
+      // Check if the invitation is resent and site managers desided to change
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/social_group_invite.module:669]: Fixed.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

File: modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    // higher than old ones.
+    $duplicate->set('created', $time);
+    $duplicate->setChangedTime($time);
+
+    // Set new invite to temp storage. This needs to identify if recipient of
+    // this invite should receive the altered invite message.
+    // Unfortunately, sending emails is triggering on entity inserting,
+    // and we don't have an "id" yet, so we should "uuid" property.
+    /* @see ginvite_group_content_insert() */
+    $values = (array) $this->tempStore->get(SocialGroupInviteResend::TEMP_STORE_ID);
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/src/Plugin/Action/SocialGroupInviteResend.php]: Fixed.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Review by dmytriikaiun — CHANGES_REQUESTED** [review]

File: modules/social_features/social_group/modules/social_group_invite/social_group_invite.module (hunk 7 above)
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_invite/social_group_invite.module]: fixed

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Review by dmytriikaiun — APPROVED** [review]

**Review by Ressinel — APPROVED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2757-kajo6lwjpzqnuv78qjjbko5bfytjwtpa.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6202bc9a046a4b8d29c1a62a

**Pull request merged**